Release lokomotive v0.4.0 (#910)
kinvolk/lokomotive: CHANGELOG.md

> ## v0.4.0 - 2020-09-04

Reviewer: Bump date ?

> ### Changes in v0.4.0

Reviewer: Release header is missing, with release name.

- Introduce platform.PostApplyHook interface and implement it for AKS cluster ([#886](https://github.com/kinvolk/lokomotive/pull/886)).

- aws-ebs-csi-driver: add NetworkPolicy allowing access to metadata ([#865](https://github.com/kinvolk/lokomotive/pull/865)).

- pkg/components/cluster-autoscaler: fix checking device uniqueness ([#768](https://github.com/kinvolk/lokomotive/pull/768)).

#### Development

- Replace use of github.com/pkg/errors.Wrapf with fmt ([#831](https://github.com/kinvolk/lokomotive/pull/831), [#877](https://github.com/kinvolk/lokomotive/pull/877)).

- Refactor assets handling ([#807](https://github.com/kinvolk/lokomotive/pull/807)).

- cli/cmd: improve --kubeconfig-file flag help message formatting ([#818](https://github.com/kinvolk/lokomotive/pull/818)).

- Use host's /etc/hosts entries for bootkube ([#409](https://github.com/kinvolk/lokomotive/pull/409)).

- Refactor Terraform executor ([#794](https://github.com/kinvolk/lokomotive/pull/794)).

- Pass kubeconfig content around rather than file path ([#631](https://github.com/kinvolk/lokomotive/pull/631)).

### Upgrading from v0.3.0

TODOs

> - Manual upgrade of the kubelet. Add link to the doc here.

Author: remove this.

#### Lokoctl Host binary upgrades

> ##### ct

Reviewer: The binary is not called `ct`, as title may suggest, but `terraform-provider-ct`.

- Update the `ct` Terraform provider to `v0.6.1`, find the install instructions [here](https://github.com/poseidon/terraform-provider-ct/blob/v0.6.1/README.md#install).

#### Configuration syntax changes

> ##### Disable TLS Bootstrap

Reviewer: How is that configuration syntax change? Shouldn't this be placed in the changelog above, where breaking change is mentioned? BTW, did we try to allow graceful upgrades? I think it should be possible if we leave old RBAC rules for the time being?

Author: A user should not forget to enable that flag, so it is placed under the general section called "Upgrading from v0.3.0".

Author: What do you mean by "graceful upgrades"?

Reviewer: Graceful upgrades, I mean cluster won't break after upgrading to v0.4.0, which is currently the case, right?

Reviewer 2: I'd use `####` and move it away from "Configuration syntax changes".

Reviewer: I'd remove this `if`: https://github.com/kinvolk/lokomotive/pull/618/commits/29bacb2b026daef284afa94012c16b371d155c9c#diff-a1a676d22e219f2a2ca28b4c8152354fR1 and then see if cluster upgrade works fine.

Author: How can the cluster upgrade be called fine if the control plane nodes is recreated?

Reviewer: They won't be. They will keep using old certificate (as RBAC rule will be in place still), until user decides to re-create them. Then they will be switched to TLS bootstrapping.

Author: > They won't be.

When you say that the nodes won't be recreated is it because of the new change: #907 ?

Should we include that change in this release? How does a user decide they want to recreate the nodes?

Reviewer: IMO, that will allow graceful migration to TLS bootstrapping, as new nodes/pools will be using it, so I would include that in the release.

>When you say that the nodes won't be recreated is it because of the new change: #907 ?

Yes.

>How does a user decide they want to recreate the nodes?

That's something we should define.

Author: If we don't define the node recreation, should we even make the change that you are suggesting? IMO this change wasn't in the milestone and should not be included.

Reviewer: We decided to require the parameter now and re-consider it as part of the next release.

Disable TLS bootstrapping to avoid recreation of cluster. Add the following attribute to the `cluster ...` block:

```tf
cluster "packet" {
  enable_tls_bootstrap = false
...
```

#### Cluster upgrade steps

> Go to your cluster's directory and run the following command.

Reviewer: ```suggestion
Go to your cluster's directory and run the following command:
```

```

You should see that `rook-version` slowly changes to `v1.4.2`.

- Watch for the Ceph version update

Run the following command to keep an eye on the Ceph version update as the new pods come up:

```
watch --exec kubectl -n rook get deployments -l rook_cluster=rook -o jsonpath='{range .items[*]}{.metadata.name}{"  \treq/upd/avl: "}{.spec.replicas}{"/"}{.status.updatedReplicas}{"/"}{.status.readyReplicas}{"  \tceph-version="}{.metadata.labels.ceph-version}{"\n"}{end}'
```

TODO:
You should see that `ceph-version` slowly changes to `15.`

- Keep an eye on the events in the rook namespace

```
kubectl -n rook get events -w
```

- Ceph Dashboard

Keep it open in one window, but sometimes it is more hassle than any help. It keeps reloading and logs you out automatically. See this on how to access the dashboard: https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/rook-ceph-storage.md#access-the-ceph-dashboard.

- Grafana dashboards

Keep an eye on the Grafana dashboard but the data here will always be old and the most reliable state of the system will come from the watch running inside toolbox pod.

- Run updates

```

> kubectl apply -f https://raw.githubusercontent.com/kinvolk/lokomotive/master/assets/charts/components/rook/templates/resources.yaml

Reviewer 3: Perhaps link this to a specific commit instead of the master branch to make sure it's future proof?

Reviewer: Ideally to tag, good point :)

lokoctl component apply rook rook-ceph
```

- Verify that the csi images are updated:

```
kubectl --namespace rook get pod -o jsonpath='{range .items[*]}{range .spec.containers[*]}{.image}{"\n"}' -l 'app in (csi-rbdplugin,csi-rbdplugin-provisioner,csi-cephfsplugin,csi-cephfsplugin-provisioner)' | sort | uniq
```

- Final checks:

Once everything is up to date then run following commands in the toolbox pod:

```
ceph status
ceph osd status
ceph df
rados df
```

> #### Openebs

Reviewer: ```suggestion
#### OpenEBS
```

        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true

        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: quay.io/openebs/m-upgrade:1.12.0
        imagePullPolicy: Always
      restartPolicy: OnFailure
---

```

Apply the Job manifest using `kubectl`. Check the logs of the pod started by the Job:

```console
$ kubectl get logs -n openebs cstor-vol-1001120-8b2h9
..
..
..
I0903 12:41:41.984635       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-e115f3f9-1666-4680-a932-d05bfd049087
I0903 12:41:41.994013       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-e115f3f9-1666-4680-a932-d05bfd049087} from 1.11.0 to 1.12.0
```

Verify that all the volumes are updated to the latest version by running the following command:

```console
$ kubectl get cstorvolume -A -o jsonpath='{.items[*].versionDetails.status.current}'
1.12.0 1.12.0 1.12.0
```

#### Upgrade other components

Other components are safe to upgrade by running the following command:

```

> lokoctl component apply <component name>

Reviewer: After upgrading MetalLB and storage workloads, other components can be updated automatically, there is no need to update one by one.
```suggestion
lokoctl cluster apply
```

Author: I am wary of doing it. I would prefer it to be done one at a time because if a complex set of events break some component you have no way of knowing what broke it? So I would ask the users to do one component at a time, verify things are fine then move to next.

Reviewer 2: Do you have a particular breakage in mind? I think this is pretty bad UX if the user needs to type all remaining components manually. `lokoctl component apply` should work, and if it doesn't we should fix it so I'd accept this change.

Author: > Do you have a particular breakage in mind?

No.

> I think this is pretty bad UX if the user needs to type all remaining components manually. `lokoctl component apply` should work, and if it doesn't we should fix it so I'd accept this change.

I agree. But also it is safer. I was testing the networking components like metallb, contour, cert-manager and I upgraded them once using command `lokoctl component apply metallb contour cert-manager` and things started failing now I wasn't sure why. Eventually I figured it is metallb only when I did an individual update. When metallb failed, nothing failed visually (like pods were fine everywhere) looking at things, but when I looked closely at the logs of metallb I figured it is failing. So such unknown scenarios I want to avoid, hence the upgrade to do it individually. If `lokoctl` would have co-ordinated upgrades making sure things are fine at every step and then moved ahead I would not hesitate in recommending it then. But now it is better to be safer.

Reviewer 2: Fair.

##### AWS

- AWS: Add support for custom taints and labels
  ([#832](https://github.com/kinvolk/lokomotive/pull/832)).

#### New Components

- Add component `experimental-istio-operator`
  ([#686](https://github.com/kinvolk/lokomotive/pull/686)).
- Add component `experimental-linkerd` ([#690](https://github.com/kinvolk/lokomotive/pull/690)).

#### Component updates

- Update etcd to `v3.4.13` ([#838](https://github.com/kinvolk/lokomotive/pull/838)).
- Update Calico to `v3.15.2` ([#841](https://github.com/kinvolk/lokomotive/pull/841)).
- Update Grafana to `7.1.4` and chart version `5.5.5`
  ([#842](https://github.com/kinvolk/lokomotive/pull/842)).
- Update Velero chart to `1.4.2` ([#830](https://github.com/kinvolk/lokomotive/pull/830)).
- Update external-dns chart to `3.3.0` ([#845](https://github.com/kinvolk/lokomotive/pull/845)).
- Update aws-ebs-csi-driver to `v0.6.0` ([#856](https://github.com/kinvolk/lokomotive/pull/856)).
- Update cluster-autoscaler to `v2` version `1.0.2`
  ([#859](https://github.com/kinvolk/lokomotive/pull/859)).
- Update cert-manager to `v0.16.1` ([#847](https://github.com/kinvolk/lokomotive/pull/847)).

> - Update openebs to `v1.12.0` ([#781](https://github.com/kinvolk/lokomotive/pull/781)).

Reviewer: We should consistently use either `component name` or Project name here I think.
```suggestion
- Update OpenEBS to `v1.12.0` ([#781](https://github.com/kinvolk/lokomotive/pull/781)).
```

> - Update metallb to `v0.1.0-789-g85b7a46a` ([#885](https://github.com/kinvolk/lokomotive/pull/885)).

Reviewer: Ditto component names + all other items on this list :smile: 
```suggestion
- Update MetalLB to `v0.1.0-789-g85b7a46a` ([#885](https://github.com/kinvolk/lokomotive/pull/885)).
```

- Update rook to `v1.4.2` ([#879](https://github.com/kinvolk/lokomotive/pull/879)).
- Use new bootkube image at version `v0.14.0-helm-7047a87`
  ([#775](https://github.com/kinvolk/lokomotive/pull/775)), later updated to `v0.14.0-helm-ec64535`
  as a part of ([#704](https://github.com/kinvolk/lokomotive/pull/704)).
- Update prometheus operator to `0.41.0` and chart version `9.3.0`
  ([#757](https://github.com/kinvolk/lokomotive/pull/757)).
- Update contour to v1.7.0 ([#771](https://github.com/kinvolk/lokomotive/pull/771)).

#### Terraform Providers Updates

- Update all Terraform providers to latest versions
  ([#835](https://github.com/kinvolk/lokomotive/pull/835)).

#### UX

- Add autocomplete for bash and zsh in lokoctl
  ([#880](https://github.com/kinvolk/lokomotive/pull/880)).

>     Now just run the following command and start using auto-completion for lokoctl.

Reviewer: I think style guide tells to avoid using `just`? 
```suggestion
    Run the following command and start using auto-completion for lokoctl.
```

  # The name can be any valid K8s string for name. This example uses
  # the following convention: cstor-spc-<flattened-from-to-versions>
  name: cstor-spc-11101120

  # VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      # VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

        # --from-version is the current version of the pool
        - "--from-version=1.11.0"

        # --to-version is the version desired upgrade version
        - "--to-version=1.12.0"

        # Bulk upgrade is supported from 1.9
        # To make use of it, please provide the list of SPCs
        # as mentioned below

>         - "cstor-pool-openebs-replica1"

Author: Does this change in wording sound ok:

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index dbfa24b0e..6894e9844 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -325,7 +325,7 @@ lokoctl component apply openebs-operator openebs-storage-class
 
 ##### Upgrade cStor Pools
 
-* Extract the SPC name using the following command:
+* Extract the SPC name using the following command and replace it in the subsequent YAML file:
 
 ```console
 $ kubectl get spc

```

>         - "cstor-pool-openebs-replica1"

Author: made the change.

>         - "cstor-pool-openebs-replica1"

Reviewer: @surajssd maybe just commit it and re-request the review? :smile:

  # The name can be any valid K8s string for name. This example uses
  # the following convention: cstor-vol-<flattened-from-to-versions>
  name: cstor-vol-11101120

  # VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return the status of the openebs-apiserver.
  namespace: openebs

spec:
  backoffLimit: 4
  template:
    spec:
      # VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-volume"

        # --from-version is the current version of the volume
        - "--from-version=1.11.0"

        # --to-version is the version desired upgrade version
        - "--to-version=1.12.0"

        # Bulk upgrade is supported from 1.9
        # To make use of it, please provide the list of cstor volumes

>         # as mentioned below

Author: made the change.

- Update Cluster Autoscaler to `v2` version `1.0.2`
  ([#859](https://github.com/kinvolk/lokomotive/pull/859)).
- Update cert-manager to `v0.16.1` ([#847](https://github.com/kinvolk/lokomotive/pull/847)).
- Update OpenEBS to `v1.12.0` ([#781](https://github.com/kinvolk/lokomotive/pull/781)).
- Update MetalLB to `v0.1.0-789-g85b7a46a` ([#885](https://github.com/kinvolk/lokomotive/pull/885)).
- Update Rook to `v1.4.2` ([#879](https://github.com/kinvolk/lokomotive/pull/879)).
- Use new bootkube image at version `v0.14.0-helm-7047a87`
  ([#775](https://github.com/kinvolk/lokomotive/pull/775)), later updated to `v0.14.0-helm-ec64535`
  as a part of ([#704](https://github.com/kinvolk/lokomotive/pull/704)).
- Update Prometheus operator to `0.41.0` and chart version `9.3.0`
  ([#757](https://github.com/kinvolk/lokomotive/pull/757)).
- Update contour to v1.7.0 ([#771](https://github.com/kinvolk/lokomotive/pull/771)).

#### Terraform Providers Updates

- Update all Terraform providers to latest versions
  ([#835](https://github.com/kinvolk/lokomotive/pull/835)).

#### UX

- Add autocomplete for bash and zsh in lokoctl
  ([#880](https://github.com/kinvolk/lokomotive/pull/880)).

>     Run the following command and start using auto-completion for lokoctl.

Reviewer: ```suggestion
    Run the following command to start using auto-completion for lokoctl:
```

Author: Actually both make sense.

  [here](https://github.com/kinvolk/lokomotive/blob/v0.4.0/docs/how-to-guides/upgrade-etcd.md).
- Manually upgrade the kubelet running on the nodes, by following the steps mentioned in the doc
  [here](https://github.com/kinvolk/lokomotive/blob/v0.4.0/docs/how-to-guides/upgrade-bootstrap-kubelet.md).

#### Manual Cluster Changes

The latest version of Metallb changes the labels of the ingress nodes. Label all the nodes that have
`asn` set with the new labels:

```
kubectl label $(kubectl get nodes -o name -l metallb.universe.tf/my-asn) \
  metallb.lokomotive.io/my-asn=65000 metallb.lokomotive.io/peer-asn=65530
```

Find a peer address of a node and assign it new label:

```bash
for node in $(kubectl get nodes -o name -l metallb.universe.tf/peer-address); do
  peer_ip=$(kubectl get $node -o jsonpath='{.metadata.labels.metallb\.universe\.tf/peer-address}')
  kubectl label $node metallb.lokomotive.io/peer-address=$peer_ip
done
```

Now it is safe to update:

```
lokoctl component apply metallb
```

#### Ceph Upgrade steps

These steps are curated from the upgrade doc provided by rook:
https://rook.io/docs/rook/master/ceph-upgrade.html.

- Keep note of the CSI images:

> ```

Reviewer 2: Sorry for being pedantic, but can you update the Ceph upgrade steps so they are correctly indented under the bullet points like we did for the UX section? :pray:

- Add component `experimental-linkerd` ([#690](https://github.com/kinvolk/lokomotive/pull/690)).

#### Component updates

- Update etcd to `v3.4.13` ([#838](https://github.com/kinvolk/lokomotive/pull/838)).

- Update Calico to `v3.15.2` ([#841](https://github.com/kinvolk/lokomotive/pull/841)).

- Update Grafana to `7.1.4` and chart version `5.5.5`
  ([#842](https://github.com/kinvolk/lokomotive/pull/842)).

- Update Velero chart to `1.4.2` ([#830](https://github.com/kinvolk/lokomotive/pull/830)).

- Update external-dns chart to `3.3.0` ([#845](https://github.com/kinvolk/lokomotive/pull/845)).

- Update aws-ebs-csi-driver to `v0.6.0` ([#856](https://github.com/kinvolk/lokomotive/pull/856)).

- Update cluster-autoscaler to `v2` version `1.0.2`
  ([#859](https://github.com/kinvolk/lokomotive/pull/859)).

- Update cert-manager to `v0.16.1` ([#847](https://github.com/kinvolk/lokomotive/pull/847)).

- Update openebs to `v1.12.0` ([#781](https://github.com/kinvolk/lokomotive/pull/781)).

- Update metallb to `v0.1.0-789-g85b7a46a` ([#885](https://github.com/kinvolk/lokomotive/pull/885)).

- Update rook to `v1.4.2` ([#879](https://github.com/kinvolk/lokomotive/pull/879)).

> - Use new bootkube image at version `v0.14.0-helm-7047a87`

Reviewer 4: We are using more latest image `v0.14.0-helm-ec64535`

Reviewer 4: https://github.com/kinvolk/lokomotive/pull/704/commits/41b63b02c758003f5df510b4361d435fff1a5910

Author: Does it make sense to call it something like this:

```md
- Use new bootkube image at version `v0.14.0-helm-7047a87`
  ([#775](https://github.com/kinvolk/lokomotive/pull/775)), later updated to `v0.14.0-helm-ec64535`
  as a part of ([#704](https://github.com/kinvolk/lokomotive/pull/704)).
```

Reviewer 4: Sounds fair

  ([#775](https://github.com/kinvolk/lokomotive/pull/775)).

- Update prometheus operator to `0.41.0` and chart version `9.3.0`
  ([#757](https://github.com/kinvolk/lokomotive/pull/757)).

- Update contour to v1.7.0 ([#771](https://github.com/kinvolk/lokomotive/pull/771)).

#### Terraform Providers Updates

- Update all Terraform providers to latest versions
  ([#835](https://github.com/kinvolk/lokomotive/pull/835)).

#### UX

- Add autocomplete for bash and zsh in lokoctl
  ([#880](https://github.com/kinvolk/lokomotive/pull/880)).

Now just run the following command and start using auto-completion for lokoctl.

```bash
source <(lokoctl completion bash)

> ```

Reviewer 4: Also, I think if we are adding an example, we should add for zsh also. Users should know we support both Bash and zsh

- Add `kubeconfig` fallback to Terraform state
  ([#701](https://github.com/kinvolk/lokomotive/pull/701)).

#### Features

- Add label `lokomotive.kinvolk.io/name: <namespace_name>` to all namespaces
  ([#646](https://github.com/kinvolk/lokomotive/pull/646)).

> - Add admission webhook to lokomotive, which disables automount of `default` service account token

Reviewer 4: ```suggestion
- Add admission webhook to lokomotive, which disables automounting `default` service account token
```

  ([#704](https://github.com/kinvolk/lokomotive/pull/704)).

> - [Breaking Change] Kubelet joins cluster using TLS Bootstrapping now, add flag

Reviewer 4: As this is breaking change, shouldn't this be in Bold?

- pkg/components/cluster-autoscaler: fix checking device uniqueness
  ([#768](https://github.com/kinvolk/lokomotive/pull/768)).

#### Development

- Replace use of github.com/pkg/errors.Wrapf with fmt
  ([#831](https://github.com/kinvolk/lokomotive/pull/831),
  [#877](https://github.com/kinvolk/lokomotive/pull/877)).

- Refactor assets handling ([#807](https://github.com/kinvolk/lokomotive/pull/807)).

- cli/cmd: improve --kubeconfig-file flag help message formatting
  ([#818](https://github.com/kinvolk/lokomotive/pull/818)).

- Use host's /etc/hosts entries for bootkube
  ([#409](https://github.com/kinvolk/lokomotive/pull/409)).

- Refactor Terraform executor ([#794](https://github.com/kinvolk/lokomotive/pull/794)).

- Pass kubeconfig content around rather than a file path
  ([#631](https://github.com/kinvolk/lokomotive/pull/631)).

### Upgrading from v0.3.0

#### Lokoctl Host binary upgrades

##### terraform-provider-ct

- Update the `ct` Terraform provider to `v0.6.1`, find the install instructions
  [here](https://github.com/poseidon/terraform-provider-ct/blob/v0.6.1/README.md#install).

#### Configuration syntax changes

##### Disable TLS Bootstrap

> Disable TLS bootstrapping to avoid recreation of cluster. Add the following attribute to the

Reviewer 2: ```suggestion
In this release we added TLS bootstrapping support by default. To avoid cluster recreation, disable it by adding the following attribute to the
```

components:

```
watch --exec kubectl -n rook get deployments -l rook_cluster=rook -o jsonpath='{range .items[*]}{.metadata.name}{"  \treq/upd/avl: "}{.spec.replicas}{"/"}{.status.updatedReplicas}{"/"}{.status.readyReplicas}{"  \trook-version="}{.metadata.labels.rook-version}{"\n"}{end}'
```

You should see that `rook-version` slowly changes to `v1.4.2`.

- Watch for the Ceph version update

Run the following command to keep an eye on the Ceph version update as the new pods come up:

```
watch --exec kubectl -n rook get deployments -l rook_cluster=rook -o jsonpath='{range .items[*]}{.metadata.name}{"  \treq/upd/avl: "}{.spec.replicas}{"/"}{.status.updatedReplicas}{"/"}{.status.readyReplicas}{"  \tceph-version="}{.metadata.labels.ceph-version}{"\n"}{end}'
```

You should see that `ceph-version` slowly changes to `15.`

- Keep an eye on the events in the rook namespace

```
kubectl -n rook get events -w
```

- Ceph Dashboard

Keep it open in one window, but sometimes it is more hassle than any help. It keeps reloading and
logs you out automatically. See this on how to access the dashboard:

> https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/rook-ceph-storage.md#access-the-ceph-dashboard.

Reviewer 2: ```suggestion
https://github.com/kinvolk/lokomotive/blob/v0.4.0/docs/how-to-guides/rook-ceph-storage.md#access-the-ceph-dashboard.
```
